File: IndiaCP_DL_Integration/src/main/java/com/barclays/indiacp/dl/integration/BorrowingLimitBoardResolutionApi.java
package com.barclays.indiacp.dl.integration;

import com.barclays.indiacp.dl.utils.DLAttachmentUtils;
import com.barclays.indiacp.dl.utils.DLUtils;
import com.barclays.indiacp.model.BoardResolutionBorrowingLimitDocument;
import org.glassfish.jersey.media.multipart.FormDataParam;

import javax.ws.rs.*;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.io.InputStream;

/**
 * Created by ritukedia on 23/12/16.
 */
@Path("boardresolution")
public class BorrowingLimitBoardResolutionApi {
    WebTarget dlRestEndPoint;
    String resourcePath;

    public BorrowingLimitBoardResolutionApi() {
        this.resourcePath = "boardresolution";
        this.dlRestEndPoint = DLUtils.getDLRestEndPoint();
    }

    @POST
    @Path("issueBoardResolution")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    public Response issueBoardResolution(
                                      @FormDataParam("boardResolutionDetails") BoardResolutionBorrowingLimitDocument brDetails,
                                      @FormDataParam("file") InputStream uploadedInputStream)
    {
        //Upload attachment to DL
        String docHash = DLAttachmentUtils.getInstance().uploadAttachment(uploadedInputStream);

        //Set the Uploaded DocHash on the Contract State to establish proof of existence of the document on the DL
        brDetails.setDocHash(docHash);

        Response response = DLUtils.getDLRestEndPoint()
                            .path(resourcePath + "/" + "issueBoardResolution").request().post(Entity.json(DLUtils.getJSONString(brDetails)));
        return response;

    }

    @POST
    @Path("amendBoardResolution")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    public Response amendBoardResolution(
            @FormDataParam("boardResolutionDetails") BoardResolutionBorrowingLimitDocument brDetails,
            @FormDataParam("file") InputStream uploadedInputStream)
    {
        //Upload attachment to DL
        String docHash = DLAttachmentUtils.getInstance().uploadAttachment(uploadedInputStream);

        //Set the Uploaded DocHash on the Contract State to establish proof of existence of the document on the DL
        brDetails.setDocHash(docHash);

        Response response = DLUtils.getDLRestEndPoint()
                .path(resourcePath + "/" + "amendBoardResolution").request().post(Entity.json(DLUtils.getJSONString(brDetails)));
        return response;

    }

    @GET
    @Path("fetchBoardResolution")
    @Produces(MediaType.APPLICATION_JSON)
    public Response fetchBoardResolution() {
        Response response = dlRestEndPoint.path(resourcePath + "/" + "fetchBoardResolution").request().get();

        return response;
    }

    @GET
    @Path("cancelBoardResolution")
    @Produces(MediaType.APPLICATION_JSON)
    public Response cancelBoardResolution() {
        Response response = dlRestEndPoint.path(resourcePath + "/" + "cancelBoardResolution").request().get();

        return response;
    }
}

File: IndiaCP_Corda/india-cp-cordapp/src/main/kotlin/com/barclays/indiacp/cordapp/api/BorrowingLimitBoardResolutionApi.kt
package com.barclays.indiacp.cordapp.api

import com.barclays.indiacp.cordapp.contract.BorrowingLimitBoardResolution
import com.barclays.indiacp.cordapp.contract.LegalEntityDocumentOwnableState
import com.barclays.indiacp.cordapp.protocol.issuer.BorrowingLimitBoardResolutionFlows
import com.barclays.indiacp.cordapp.utilities.ErrorUtils
import com.barclays.indiacp.cordapp.utilities.ModelUtils
import com.barclays.indiacp.model.*
import net.corda.core.node.ServiceHub
import net.corda.core.utilities.Emoji
import net.corda.core.utilities.loggerFor
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * This is the REST Client API for Managing the Borrowing Limit Board Resolution. This is part of the reference data but
 * is required in the Commercial Paper process and is legally certified by the IPA.
 *
 * Created by ritukedia
 */
@Path("boardresolution")
class BorrowingLimitBoardResolutionApi(val services: ServiceHub){

    private companion object {
        val logger = loggerFor<BorrowingLimitBoardResolutionApi>()
    }

    @POST
    @Path("issueBoardResolution")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    fun issueBoardResolution(boardResolutionDocument: BoardResolutionBorrowingLimitDocument): Response
    {
        try {
            val boardResolutionContractState : LegalEntityDocumentOwnableState = ModelUtils.boardResolutionStateFromModel(boardResolutionDocument, services)
            val stx = services.invokeFlowAsync(BorrowingLimitBoardResolutionFlows::class.java, boardResolutionContractState, BorrowingLimitBoardResolution.Commands.Issue::class.java.simpleName).resultFuture.get()
            logger.info("Issued Credit Rating with CR Document\n\nFinal transaction is:\n\n${Emoji.renderIfSupported(stx.tx)}")

            return Response.status(Response.Status.OK).entity(getBoardResolutionContractState()).build()

        } catch (ex: Throwable) {
            logger.info("Exception when creating credit rating: ${ex.toString()}")
            return ErrorUtils.errorHttpResponse(ex, errorCode = CreditRatingError.CREATION_ERROR)
        }
    }

    @POST
    @Path("amendBoardResolution")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    fun amendBoardResolution(boardResolutionDocument: BoardResolutionBorrowingLimitDocument): Response
    {
        try {
            val boardResolutionContractState : LegalEntityDocumentOwnableState = ModelUtils.boardResolutionStateFromModel(boardResolutionDocument, services)
            val stx = services.invokeFlowAsync(BorrowingLimitBoardResolutionFlows::class.java, boardResolutionContractState, BorrowingLimitBoardResolution.Commands.Amend::class.java.simpleName).resultFuture.get()
            logger.info("Amended Credit Rating with CR Document\n\nFinal transaction is:\n\n${Emoji.renderIfSupported(stx.tx)}")

            return Response.status(Response.Status.OK).entity(getBoardResolutionContractState()).build()

        } catch (ex: Throwable) {
            logger.info("Exception when creating credit rating: ${ex.toString()}")
            return ErrorUtils.errorHttpResponse(ex, errorCode = CreditRatingError.AMENDMENT_ERROR)
        }

    }

    @GET
    @Path("cancelBoardResolution")
    @Produces(MediaType.APPLICATION_JSON)
    fun cancelBoardResolution(): Response
    {
        try {
            val stx = services.invokeFlowAsync(BorrowingLimitBoardResolutionFlows::class.java, null, BorrowingLimitBoardResolution.Commands.Cancel::class.java.simpleName).resultFuture.get()
            logger.info("Cancelled Board Resolution \n\nFinal transaction is:\n\n${Emoji.renderIfSupported(stx.tx)}")

            return Response.status(Response.Status.OK).entity("Board Resolution Document for this Legal Entity has been Cancelled").build()

        } catch (ex: Throwable) {
            logger.info("Exception when creating credit rating: ${ex.toString()}")
            return ErrorUtils.errorHttpResponse(ex, errorCode = CreditRatingError.CANCELLATION_ERROR)
        }
    }

    @GET
    @Path("fetchBoardResolution")
    @Produces(MediaType.APPLICATION_JSON)
    fun fetchBoardResolution(): Response
    {
        try {
            return Response.status(Response.Status.OK).entity(getBoardResolutionContractState() ?: "No Board Resolution Documents Uploaded").build()
        } catch (ex: Throwable) {
            logger.info("Exception when fetching board resolution: ${ex.toString()}")
            return ErrorUtils.errorHttpResponse(ex, errorCode = CreditRatingError.FETCH_ERROR)
        }
    }

    fun getBoardResolutionModel(): BoardResolutionBorrowingLimitDocument? {
        val boardResolutionContractState = getBoardResolutionContractState()
        if (boardResolutionContractState != null)
            return ModelUtils.boardResolutionModelFromState(boardResolutionContractState)
        else
            return null
    }

    fun getBoardResolutionContractState(): BorrowingLimitBoardResolution.State? {
        val states = services.vaultService.currentVault.statesOfType<BorrowingLimitBoardResolution.State>()
        return if (states.isEmpty()) null else {
            val datas = states.map { it.state.data }
            return if (datas.isEmpty()) null else datas[0]
        }
    }

}
